PR: https://github.com/odoo/odoo/pull/214615

From: 1ee02f8a47d42d3ba3fd11ffcf8d9768ea17678e
From: Stephane Vanmeerhaeghe (stva)
Date: 2025-07-03 19:19:28

Breaking data model changes scores: del:1 + add:0, change matches:
-    _inherit = 'ir.module.module'

Total Changes: 470

[FIX] point_of_sale, pos_*: Fix loading behavior between self and POS

*: l10n_ar_pos, l10n_be_pos_sale, l10n_es_edi_tbai_pos, l10n_in_pos,
l10n_pe_pos, pos_discount, pos_event, pos_hr, pos_loyalty,
pos_online_payment_self_order, pos_restaurant, pos_sale, pos_self_order

Before this commit, the loading data behavior between self-ordering and
PoS was merged but not consistent. This led to issues when we try to
add some data to load for both self-ordering and PoS. At the moment, we
need to create two separate method to handle both cases.

Now with this commit, we have consistent methods which will load data
for both self-ordering and PoS. This will allow us to have a more
consistent behavior and avoid issues when adding data to load.

An explicit example is about `_server_version` object which is added to
the `pos.session` model and used when printing the receipt. Since
the session isn't always available in self-ordering, sometimes that
object isn't available, leading to issues when printing the receipt.

closes odoo/odoo#214615

Related: odoo/enterprise#87883
Signed-off-by: David Monnom (moda) <moda@odoo.com>

================================= pseudo patch: =================================

--- a/addons/point_of_sale/models/account_fiscal_position.py
+++ b/addons/point_of_sale/models/account_fiscal_position.py
@@ -6,11 +6,11 @@ class AccountFiscalPosition(models.Model):
     _inherit = ['account.fiscal.position', 'pos.load.mixin']
 
     @api.model
-    def _load_pos_data_domain(self, data):
+    def _load_pos_data_domain(self, data, config):
         fp_ids = [preset['fiscal_position_id'] for preset in data['pos.preset']]
         partner_fp_ids = list({partner['fiscal_position_id'] for partner in data['res.partner'] if partner['fiscal_position_id']}) if 'res.partner' in data.keys() else []
-        return [('id', 'in', data['pos.config'][0]['fiscal_position_ids'] + fp_ids + partner_fp_ids)]
+        return [('id', 'in', config.fiscal_position_ids.ids + fp_ids + partner_fp_ids)]
 
     @api.model
-    def _load_pos_data_fields(self, config_id):
+    def _load_pos_data_fields(self, config):
         return ['id', 'name', 'display_name', 'tax_map']

--- a/addons/point_of_sale/models/account_move.py
+++ b/addons/point_of_sale/models/account_move.py
@@ -101,8 +101,8 @@ class AccountMove(models.Model):
         return action
 
     @api.model
-    def _load_pos_data_fields(self, config_id):
-        result = super()._load_pos_data_fields(config_id)
+    def _load_pos_data_fields(self, config):
+        result = super()._load_pos_data_fields(config)
         return result or ['id']
 
 class AccountMoveLine(models.Model):

--- a/addons/point_of_sale/models/account_tax.py
+++ b/addons/point_of_sale/models/account_tax.py
@@ -52,11 +52,11 @@ class AccountTax(models.Model):
         return used_taxes
 
     @api.model
-    def _load_pos_data_domain(self, data):
-        return self.env['account.tax']._check_company_domain(data['pos.config'][0]['company_id'])
+    def _load_pos_data_domain(self, data, config):
+        return self.env['account.tax']._check_company_domain(config.company_id.id)
 
     @api.model
-    def _load_pos_data_fields(self, config_id):
+    def _load_pos_data_fields(self, config):
         return [
             'id', 'name', 'price_include', 'include_base_amount', 'is_base_affected', 'has_negative_factor',
             'amount_type', 'children_tax_ids', 'amount', 'company_id', 'id', 'sequence', 'tax_group_id',

--- a/addons/point_of_sale/models/account_tax_group.py
+++ b/addons/point_of_sale/models/account_tax_group.py
@@ -6,10 +6,10 @@ class AccountTaxGroup(models.Model):
     _inherit = ['account.tax.group', 'pos.load.mixin']
 
     @api.model
-    def _load_pos_data_domain(self, data):
+    def _load_pos_data_domain(self, data, config):
         tax_group_ids = [tax_data['tax_group_id'] for tax_data in data['account.tax']]
         return [('id', 'in', tax_group_ids)]
 
     @api.model
-    def _load_pos_data_fields(self, config_id):
+    def _load_pos_data_fields(self, config):
         return ['id', 'name', 'pos_receipt_label']

--- a/addons/point_of_sale/models/decimal_precision.py
+++ b/addons/point_of_sale/models/decimal_precision.py
@@ -6,5 +6,5 @@ class DecimalPrecision(models.Model):
     _inherit = ['decimal.precision', 'pos.load.mixin']
 
     @api.model
-    def _load_pos_data_fields(self, config_id):
+    def _load_pos_data_fields(self, config):
         return ['id', 'name', 'digits']

--- a/addons/point_of_sale/models/ir_binary.py
+++ b/addons/point_of_sale/models/ir_binary.py

--- a/addons/point_of_sale/models/ir_module_module.py
+++ b/addons/point_of_sale/models/ir_module_module.py
@@ -3,24 +3,12 @@ from odoo import api, models
 
 class IrModuleModule(models.Model):
     _name = 'ir.module.module'
-    _inherit = 'ir.module.module'
+    _inherit = ['pos.load.mixin', 'ir.module.module']
 
     @api.model
-    def _load_pos_data_fields(self, config_id):
+    def _load_pos_data_fields(self, config):
         return ['id', 'name', 'state']
 
     @api.model
-    def _load_pos_data_domain(self):
+    def _load_pos_data_domain(self, data, config):
         return [('name', '=', 'pos_settle_due')]
-
-    def _load_pos_data(self, data):
-        domain = self._load_pos_data_domain()
-        fields = self._load_pos_data_fields(data['pos.config'][0]['id'])
-        return self.search_read(domain, fields, load=False)
-
-    def _post_read_pos_data(self, data):
-        return data
-
-    def _read_pos_record(self, ids, config_id):
-        fields = self._load_pos_data_fields(self.id)
-        return self.browse(ids).read(fields, load=False)

--- a/addons/point_of_sale/models/pos_bill.py
+++ b/addons/point_of_sale/models/pos_bill.py
@@ -23,9 +23,9 @@ class PosBill(models.Model):
         return result.id, result.display_name
 
     @api.model
-    def _load_pos_data_domain(self, data):
-        return ['|', ('id', 'in', data['pos.config'][0]['default_bill_ids']), ('pos_config_ids', '=', False)]
+    def _load_pos_data_domain(self, data, config):
+        return ['|', ('id', 'in', config.default_bill_ids.ids), ('pos_config_ids', '=', False)]
 
     @api.model
-    def _load_pos_data_fields(self, config_id):
+    def _load_pos_data_fields(self, config):
         return ['id', 'name', 'value']

--- a/addons/point_of_sale/models/pos_category.py
+++ b/addons/point_of_sale/models/pos_category.py
@@ -37,17 +37,15 @@ class PosCategory(models.Model):
     has_image = fields.Boolean(compute='_compute_has_image')
 
     @api.model
-    def _load_pos_data_domain(self, data):
+    def _load_pos_data_domain(self, data, config):
         domain = []
-        limited_categories = data['pos.config'][0]['limit_categories']
-        if limited_categories:
-            available_category_ids = data['pos.config'][0]['iface_available_categ_ids']
-            category_ids = self.env['pos.category'].browse(available_category_ids)._get_descendants().ids
+        if config.limit_categories:
+            category_ids = config.iface_available_categ_ids._get_descendants().ids
             domain += [('id', 'in', category_ids)]
         return domain
 
     @api.model
-    def _load_pos_data_fields(self, config_id):
+    def _load_pos_data_fields(self, config):
         return ['id', 'name', 'parent_id', 'child_ids', 'write_date', 'has_image', 'color', 'sequence', 'hour_until', 'hour_after']
 
     def _get_hierarchy(self) -> List[str]:

--- a/addons/point_of_sale/models/pos_config.py
+++ b/addons/point_of_sale/models/pos_config.py
@@ -10,7 +10,7 @@ from odoo import api, fields, models, _, Command, tools, SUPERUSER_ID
 from odoo.http import request
 from odoo.exceptions import AccessError, ValidationError, UserError
 from odoo.tools import SQL, convert
-from odoo.osv import expression
+from odoo.service.common import exp_version
 
 DEFAULT_LIMIT_LOAD_PRODUCT = 5000
 DEFAULT_LIMIT_LOAD_PARTNER = 100
@@ -207,7 +207,8 @@ class PosConfig(models.Model):
         static_records = {}
 
         for model, ids in records.items():
-            static_records[model] = self.env[model]._read_pos_record(ids, self.id)
+            records = self.env[model].browse(ids)
+            static_records[model] = self.env[model]._load_pos_data_read(records, self)
 
         self._notify('SYNCHRONISATION', {
             'static_records': static_records,
@@ -226,10 +227,10 @@ class PosConfig(models.Model):
             dynamic_records[model] = self.env[model].search(domain)
 
         pos_order_data = dynamic_records.get('pos.order') or self.env['pos.order']
-        data = pos_order_data.read_pos_data([], self.id)
+        data = pos_order_data.read_pos_data([], self)
 
         for key, records in dynamic_records.items():
-            fields = self.env[key]._load_pos_data_fields(self.id)
+            fields = self.env[key]._load_pos_data_fields(self)
             ids = list(set(records.ids + [record['id'] for record in data.get(key, [])]))
             dynamic_records[key] = self.env[key].browse(ids).read(fields, load=False)
 
@@ -243,20 +244,31 @@ class PosConfig(models.Model):
         }
 
     @api.model
-    def _load_pos_data_domain(self, data):
-        return [('id', '=', data['pos.session'][0]['config_id'])]
-
-    def _load_pos_data(self, data):
-        domain = self._load_pos_data_domain(data)
-        fields = self._load_pos_data_fields(self.id)
-        return self.search_read(domain, fields, load=False)
-
-    def _post_read_pos_data(self, data):
-        if not data[0]['use_pricelist']:
-            data[0]['pricelist_id'] = False
-        if data:
-            data[0]['_IS_VAT'] = self.env.company.country_id.id in self.env.ref("base.europe").country_ids.ids
-        return super()._post_read_pos_data(data)
+    def _load_pos_data_domain(self, data, config):
+        return [('id', '=', config.id)]
+
+    @api.model
+    def _load_pos_data_read(self, records, config):
+        read_records = super()._load_pos_data_read(records, config)
+        record = read_records[0]
+
+        record['_server_version'] = exp_version()
+        record['_base_url'] = self.get_base_url()
+        record['_data_server_date'] = self.env.context.get('pos_last_server_date') or self.env.cr.now()
+        record['_has_cash_move_perm'] = self.env.user.has_group('account.group_account_invoice')
+        record['_pos_special_products_ids'] = self.env['pos.config']._get_special_products().ids
+
+        # Add custom fields for 'formula' taxes.
+        # We can ignore data for _load_pos_data_domain since isn't needed in the domain computation of account.tax
+        taxes = self.env['account.tax'].search(self.env['account.tax']._load_pos_data_domain({}, config))
+        product_fields = taxes._eval_taxes_computation_prepare_product_fields()
+        record['_product_default_values'] = \
+            self.env['account.tax']._eval_taxes_computation_prepare_product_default_values(product_fields)
+
+        if not record['use_pricelist']:
+            record['pricelist_id'] = False
+        record['_IS_VAT'] = self.env.company.country_id.id in self.env.ref("base.europe").country_ids.ids
+        return read_records
 
     @api.depends('payment_method_ids')
     def _compute_cash_control(self):

--- a/addons/point_of_sale/models/pos_load_mixin.py
+++ b/addons/point_of_sale/models/pos_load_mixin.py
@@ -8,36 +8,52 @@ class PosLoadMixin(models.AbstractModel):
     _description = "PoS data loading mixin"
 
     @api.model
-    def _load_pos_data_fields(self, config_id):
-        return []
+    def _load_pos_data_search_read(self, data, config):
+        """ Search and return records to be loaded in the pos """
+        if not config:
+            raise ValueError("config must be provided to search for PoS data.")
+
+        domain = self._server_date_to_domain(self._load_pos_data_domain(data, config))
+        if domain is False:
+            return []
+
+        records = self.search(domain)
+        return self._load_pos_data_read(records, config)
 
     @api.model
-    def _load_pos_data_domain(self, data):
+    def _load_pos_data_domain(self, data, config):
+        """ Return the domain used to filter records """
         return []
 
-    # If you need to adapt the data to be loaded in the PoS, you can
-    # override this method in your model.
-    def _load_pos_data(self, data):
-        domain = self._server_date_to_domain(self._load_pos_data_domain(data))
-        fields = self._load_pos_data_fields(data['pos.config'][0]['id'])
-        return self.search_read(domain, fields, load=False) if domain is not False else []
-
+    @api.model
     def _server_date_to_domain(self, domain):
+        """ Optionally restrict the domain to records modified after the last server sync """
+        if domain is False:
+            return domain
+
         last_server_date = self.env.context.get('pos_last_server_date', False)
         limited_loading = self.env.context.get('pos_limited_loading', True)
-        if last_server_date and domain is not False and limited_loading:
+        model_included = self._name not in ['pos.session', 'pos.config']
+
+        if limited_loading and last_server_date and model_included:
             domain = AND([domain, [('write_date', '>', last_server_date)]])
+
         return domain
 
-    def _unrelevant_records(self):
-        return self.filtered(lambda record: not record.active).ids
+    @api.model
+    def _load_pos_data_read(self, records, config):
+        """ Read specific fields from the given records """
+        if not config:
+            raise ValueError("config must be provided to read PoS data.")
 
-    def _post_read_pos_data(self, data):
-        return data
+        fields = self._load_pos_data_fields(config)
+        records = records.read(fields, load=False)
+        return records or []
 
-    def _read_pos_record(self, ids, config_id):
-        fields = self._load_pos_data_fields(self.id)
-        record_ids = self.browse(ids).exists()
-        if not record_ids:
-            return []
-        return self.with_context(config_id=config_id)._post_read_pos_data(record_ids.read(fields, load=False))
+    def _unrelevant_records(self, config):
+        return self.filtered(lambda record: not record.active).ids
+
+    @api.model
+    def _load_pos_data_fields(self, config):
+        """ Return the list of fields to be loaded """
+        return []

--- a/addons/point_of_sale/models/pos_note.py
+++ b/addons/point_of_sale/models/pos_note.py
@@ -19,9 +19,9 @@ class PosNote(models.Model):
     )
 
     @api.model
-    def _load_pos_data_domain(self, data):
-        return [('id', 'in', data['pos.config'][0]['note_ids'])] if data['pos.config'][0]['note_ids'] else []
+    def _load_pos_data_domain(self, data, config):
+        return [('id', 'in', config.note_ids.ids)] if config.note_ids else []
 
     @api.model
-    def _load_pos_data_fields(self, config_id):
+    def _load_pos_data_fields(self, config):
         return ['name', 'color']

--- a/addons/point_of_sale/models/pos_order.py
+++ b/addons/point_of_sale/models/pos_order.py
@@ -55,8 +55,8 @@ class PosOrder(models.Model):
         raise UserError(_('No open session available. Please open a new session to capture the order.'))
 
     @api.model
-    def _load_pos_data_domain(self, data):
-        return [('state', '=', 'draft'), ('config_id', '=', data['pos.config'][0]['id'])]
+    def _load_pos_data_domain(self, data, config):
+        return [('state', '=', 'draft'), ('config_id', '=', config.id)]
 
     @api.model
     def _process_order(self, order, existing_order):
@@ -1079,7 +1079,7 @@ class PosOrder(models.Model):
         next_days_orders.session_id = False
         today_orders.write({'state': 'cancel'})
         return {
-            'pos.order': today_orders.read(self._load_pos_data_fields(self.config_id.ids[0]), load=False)
+            'pos.order': self._load_pos_data_read(today_orders, self.config_id)
         }
 
     def _get_open_order(self, order):
@@ -1131,7 +1131,7 @@ class PosOrder(models.Model):
 
         # Sometime pos_orders_ids can be empty.
         pos_order_ids = self.env['pos.order'].browse(order_ids)
-        config_id = pos_order_ids.config_id.ids[0] if pos_order_ids else False
+        config = pos_order_ids.config_id[0] if pos_order_ids else False
 
         for order in pos_order_ids:
             order._ensure_access_token()
@@ -1139,25 +1139,23 @@ class PosOrder(models.Model):
                 order.config_id.notify_synchronisation(order.config_id.current_session_id.id, self.env.context.get('login_number', 0))
 
         _logger.info("PoS synchronisation #%d finished", sync_token)
-        return pos_order_ids.read_pos_data(orders, config_id)
+        return pos_order_ids.read_pos_data(orders, config)
 
     @api.model
     def read_pos_data_uuid(self, uuid):
         order = self.search([('uuid', '=', uuid)], limit=1)
         if not order:
             return {}
-        return order.read_pos_data([], order.config_id.id)
-
-    def read_pos_data(self, data, config_id):
-        # If the previous session is closed, the order will get a new session_id due to _get_valid_session in _process_order
+        return order.read_pos_data([], order.config_id)
 
+    def read_pos_data(self, data, config):
         return {
-            'pos.order': self.read(self._load_pos_data_fields(config_id), load=False) if config_id else [],
+            'pos.order': self._load_pos_data_read(self, config) if config else [],
             'pos.session': [],
-            'pos.payment': self.payment_ids.read(self.payment_ids._load_pos_data_fields(config_id), load=False) if config_id else [],
-            'pos.order.line': self.lines.read(self.lines._load_pos_data_fields(config_id), load=False) if config_id else [],
-            'pos.pack.operation.lot': self.lines.pack_lot_ids.read(self.lines.pack_lot_ids._load_pos_data_fields(config_id), load=False) if config_id else [],
-            "product.attribute.custom.value": self.lines.custom_attribute_value_ids.read(self.lines.custom_attribute_value_ids._load_pos_data_fields(config_id), load=False) if config_id else [],
+            'pos.payment': self.env['pos.payment']._load_pos_data_read(self.payment_ids, config) if config else [],
+            'pos.order.line': self.env['pos.order.line']._load_pos_data_read(self.lines, config) if config else [],
+            'pos.pack.operation.lot': self.env['pos.pack.operation.lot']._load_pos_data_read(self.lines.pack_lot_ids, config) if config else [],
+            'product.attribute.custom.value': self.env['product.attribute.custom.value']._load_pos_data_read(self.lines.custom_attribute_value_ids, config) if config else [],
         }
 
     @api.model
@@ -1433,11 +1431,11 @@ class PosOrderLine(models.Model):
     extra_tax_data = fields.Json()
 
     @api.model
-    def _load_pos_data_domain(self, data):
+    def _load_pos_data_domain(self, data, config):
         return [('order_id', 'in', [order['id'] for order in data['pos.order']])]
 
     @api.model
-    def _load_pos_data_fields(self, config_id):
+    def _load_pos_data_fields(self, config):
         return [
             'qty', 'attribute_value_ids', 'custom_attribute_value_ids', 'price_unit',
             'uuid', 'price_subtotal', 'price_subtotal_incl', 'order_id', 'note', 'price_type',
@@ -1780,11 +1778,11 @@ class PosPackOperationLot(models.Model):
     product_id = fields.Many2one('product.product', related='pos_order_line_id.product_id', readonly=False)
 
     @api.model
-    def _load_pos_data_domain(self, data):
+    def _load_pos_data_domain(self, data, config):
         return [('pos_order_line_id', 'in', [line['id'] for line in data['pos.order.line']])]
 
     @api.model
-    def _load_pos_data_fields(self, config_id):
+    def _load_pos_data_fields(self, config):
         return ['lot_name', 'pos_order_line_id', 'write_date']
 
 
@@ -1800,9 +1798,9 @@ class AccountCashRounding(models.Model):
                 _("You are not allowed to change the cash rounding configuration while a pos session using it is already opened."))
 
     @api.model
-    def _load_pos_data_domain(self, data):
-        return [('id', '=', data['pos.config'][0]['rounding_method'])]
+    def _load_pos_data_domain(self, data, config):
+        return [('id', '=', config.rounding_method.id)]
 
     @api.model
-    def _load_pos_data_fields(self, config_id):
+    def _load_pos_data_fields(self, config):
         return ['id', 'name', 'rounding', 'rounding_method', 'strategy']

--- a/addons/point_of_sale/models/pos_payment.py
+++ b/addons/point_of_sale/models/pos_payment.py
@@ -44,7 +44,7 @@ class PosPayment(models.Model):
     uuid = fields.Char(string='Uuid', readonly=True, default=lambda self: str(uuid4()), copy=False)
 
     @api.model
-    def _load_pos_data_domain(self, data):
+    def _load_pos_data_domain(self, data, config):
         return [('pos_order_id', 'in', [order['id'] for order in data['pos.order']])]
 
     @api.depends('amount', 'currency_id')

--- a/addons/point_of_sale/models/pos_payment_method.py
+++ b/addons/point_of_sale/models/pos_payment_method.py
@@ -70,11 +70,11 @@ class PosPaymentMethod(models.Model):
         }
 
     @api.model
-    def _load_pos_data_domain(self, data):
+    def _load_pos_data_domain(self, data, config):
         return ['|', ('active', '=', False), ('active', '=', True)]
 
     @api.model
-    def _load_pos_data_fields(self, config_id):
+    def _load_pos_data_fields(self, config):
         return ['id', 'name', 'is_cash_count', 'use_payment_terminal', 'split_transactions', 'type', 'image', 'sequence', 'payment_method_type', 'default_qr']
 
     @api.depends('type', 'payment_method_type')

--- a/addons/point_of_sale/models/pos_preset.py
+++ b/addons/point_of_sale/models/pos_preset.py
@@ -36,12 +36,12 @@ class PosPreset(models.Model):
                     raise ValidationError(_('The start time must be before the end time.'))
 
     @api.model
-    def _load_pos_data_domain(self, data):
-        preset_ids = data['pos.config'][0]['available_preset_ids'] + [data['pos.config'][0]['default_preset_id']]
+    def _load_pos_data_domain(self, data, config):
+        preset_ids = config.available_preset_ids.ids + [config.default_preset_id.id]
         return [('id', 'in', preset_ids)]
 
     @api.model
-    def _load_pos_data_fields(self, config_id):
+    def _load_pos_data_fields(self, config):
         return ['id', 'name', 'pricelist_id', 'fiscal_position_id', 'is_return', 'color', 'has_image', 'write_date', 'identification',
             'use_timing', 'slots_per_interval', 'interval_time', 'attendance_ids']
 

--- a/addons/point_of_sale/models/pos_printer.py
+++ b/addons/point_of_sale/models/pos_printer.py
@@ -19,9 +19,9 @@ class PosPrinter(models.Model):
     pos_config_ids = fields.Many2many('pos.config', 'pos_config_printer_rel', 'printer_id', 'config_id')
 
     @api.model
-    def _load_pos_data_domain(self, data):
-        return [('id', 'in', data['pos.config'][0]['printer_ids'])]
+    def _load_pos_data_domain(self, data, config):
+        return [('id', 'in', config.printer_ids.ids)]
 
     @api.model
-    def _load_pos_data_fields(self, config_id):
+    def _load_pos_data_fields(self, config):
         return ['id', 'name', 'proxy_ip', 'product_categories_ids', 'printer_type']

--- a/addons/point_of_sale/models/pos_session.py
+++ b/addons/point_of_sale/models/pos_session.py
@@ -8,7 +8,6 @@ from odoo import api, fields, models, _, Command
 from odoo.exceptions import AccessError, UserError, ValidationError
 from odoo.tools import float_is_zero, float_compare, plaintext2html, split_every
 from odoo.tools.constants import PREFETCH_MAX
-from odoo.service.common import exp_version
 from odoo.osv.expression import AND
 
 
@@ -134,7 +133,7 @@ class PosSession(models.Model):
         return relations
 
     @api.model
-    def _load_pos_data_models(self, config_id):
+    def _load_pos_data_models(self, config):
         return ['pos.config', 'pos.preset', 'resource.calendar.attendance', 'pos.order', 'pos.order.line', 'pos.pack.operation.lot', 'pos.payment', 'pos.payment.method', 'pos.printer',
             'pos.category', 'pos.bill', 'res.company', 'account.tax', 'account.tax.group', 'product.template', 'product.product', 'product.attribute', 'product.attribute.custom.value',
             'product.template.attribute.line', 'product.template.attribute.value', 'product.template.attribute.exclusion', 'product.combo', 'product.combo.item', 'res.users', 'res.partner', 'product.uom',
@@ -142,43 +141,26 @@ class PosSession(models.Model):
             'account.cash.rounding', 'account.fiscal.position', 'stock.picking.type', 'res.currency', 'pos.note', 'product.tag', 'ir.module.module']
 
     @api.model
-    def _load_pos_data_domain(self, data):
+    def _load_pos_data_domain(self, data, config):
         return [('id', '=', self.id)]
 
     @api.model
-    def _load_pos_data_fields(self, config_id):
+    def _load_pos_data_fields(self, config):
         return [
             'id', 'name', 'user_id', 'config_id', 'start_at', 'stop_at',
             'payment_method_ids', 'state', 'update_stock_at_closing', 'cash_register_balance_start', 'access_token'
         ]
 
-    def _load_pos_data(self, data):
-        domain = self._load_pos_data_domain(data)
-        fields = self._load_pos_data_fields(self.config_id.id)
-        data = self.search_read(domain, fields, load=False, limit=1)
-        return data
-
-    def _post_read_pos_data(self, data):
-        server_date = self.env.context.get('pos_last_server_date')
-        data[0]['_partner_commercial_fields'] = self.env['res.partner']._commercial_fields()
-        data[0]['_server_version'] = exp_version()
-        data[0]['_base_url'] = self.get_base_url()
-        data[0]['_data_server_date'] = server_date or self.env.cr.now()
-        data[0]['_has_cash_move_perm'] = self.env.user.has_group('account.group_account_invoice')
-        data[0]['_has_available_products'] = self._pos_has_valid_product()
-        data[0]['_pos_special_products_ids'] = self.env['pos.config']._get_special_products().ids
-        return super()._post_read_pos_data(data)
-
     def load_data(self, models_to_load):
         response = {}
-        response['pos.session'] = self._post_read_pos_data(self._load_pos_data(response))
+        response['pos.session'] = self._load_pos_data_search_read(response, self.config_id)
 
-        for model in self._load_pos_data_models(self.config_id.id):
+        for model in self._load_pos_data_models(self.config_id):
             if models_to_load and model not in models_to_load:
                 continue
 
             try:
-                response[model] = self.env[model].with_context(config_id=self.config_id.id)._post_read_pos_data(self.env[model]._load_pos_data(response))
+                response[model] = self.env[model]._load_pos_data_search_read(response, self.config_id)
             except AccessError:
                 response[model] = []
 
@@ -186,14 +168,14 @@ class PosSession(models.Model):
 
     def load_data_params(self):
         response = {}
-        fields = self._load_pos_data_fields(self.config_id.id)
+        fields = self._load_pos_data_fields(self.config_id)
         response['pos.session'] = {
             'fields': fields,
             'relations': self._load_pos_data_relations('pos.session', fields)
         }
 
-        for model in self._load_pos_data_models(self.config_id.id):
-            fields = self.env[model]._load_pos_data_fields(self.config_id.id)
+        for model in self._load_pos_data_models(self.config_id):
+            fields = self.env[model]._load_pos_data_fields(self.config_id)
             response[model] = {
                 'fields': fields,
                 'relations': self._load_pos_data_relations(model, fields)
@@ -207,7 +189,7 @@ class PosSession(models.Model):
             existing_records = self.env[model].browse(ids).exists()
 
             non_existent_ids = set(ids) - set(existing_records.ids)
-            inactive_ids = set(existing_records.with_context(config_id=self.config_id.id)._unrelevant_records())
+            inactive_ids = set(existing_records._unrelevant_records(self.config_id))
 
             response[model] = list(non_existent_ids | inactive_ids)
         return response
@@ -226,8 +208,9 @@ class PosSession(models.Model):
         }
 
     def get_pos_ui_product_pricelist_item_by_product(self, product_tmpl_ids, product_ids, config_id):
-        pricelist_fields = self.env['product.pricelist']._load_pos_data_fields(config_id)
-        pricelist_item_fields = self.env['product.pricelist.item']._load_pos_data_fields(config_id)
+        pos_config = self.env['pos.config'].browse(config_id)
+        pricelist_fields = self.env['product.pricelist']._load_pos_data_fields(pos_config)
+        pricelist_item_fields = self.env['product.pricelist.item']._load_pos_data_fields(pos_config)
         today = fields.Date.today()
         pricelist_item_domain = [
             '&',

--- a/addons/point_of_sale/models/product_attribute.py
+++ b/addons/point_of_sale/models/product_attribute.py
@@ -7,7 +7,7 @@ class ProductAttribute(models.Model):
     _inherit = ['product.attribute', 'pos.load.mixin']
 
     @api.model
-    def _load_pos_data_fields(self, config_id):
+    def _load_pos_data_fields(self, config):
         return ['name', 'display_type', 'template_value_ids', 'attribute_line_ids', 'create_variant']
 
 
@@ -18,11 +18,11 @@ class ProductAttributeCustomValue(models.Model):
     pos_order_line_id = fields.Many2one('pos.order.line', string="PoS Order Line", ondelete='cascade', index='btree_not_null')
 
     @api.model
-    def _load_pos_data_domain(self, data):
+    def _load_pos_data_domain(self, data, config):
         return [('pos_order_line_id', 'in', [line['id'] for line in data['pos.order.line']])]
 
     @api.model
-    def _load_pos_data_fields(self, config_id):
+    def _load_pos_data_fields(self, config):
         return ['custom_value', 'custom_product_template_attribute_value_id', 'pos_order_line_id', 'write_date']
 
 
@@ -31,11 +31,11 @@ class ProductTemplateAttributeLine(models.Model):
     _inherit = ['product.template.attribute.line', 'pos.load.mixin']
 
     @api.model
-    def _load_pos_data_fields(self, config_id):
+    def _load_pos_data_fields(self, config):
         return ['display_name', 'attribute_id', 'product_template_value_ids']
 
     @api.model
-    def _load_pos_data_domain(self, data):
+    def _load_pos_data_domain(self, data, config):
         loaded_product_tmpl_ids = list({p['id'] for p in data['product.template']})
         return [('product_tmpl_id', 'in', loaded_product_tmpl_ids)]
 
@@ -45,7 +45,7 @@ class ProductTemplateAttributeValue(models.Model):
     _inherit = ['product.template.attribute.value', 'pos.load.mixin']
 
     @api.model
-    def _load_pos_data_domain(self, data):
+    def _load_pos_data_domain(self, data, config):
         ptav_ids = {ptav_id for p in data['product.product'] for ptav_id in p['product_template_variant_value_ids']}
         ptav_ids.update({ptav_id for ptal in data['product.template.attribute.line'] for ptav_id in ptal['product_template_value_ids']})
         return AND([
@@ -55,7 +55,7 @@ class ProductTemplateAttributeValue(models.Model):
         ])
 
     @api.model
-    def _load_pos_data_fields(self, config_id):
+    def _load_pos_data_fields(self, config):
         return ['attribute_id', 'attribute_line_id', 'product_attribute_value_id', 'price_extra', 'name', 'is_custom', 'html_color', 'image', 'exclude_for']
 
 class ProductTemplateAttributeExclusion(models.Model):
@@ -63,10 +63,10 @@ class ProductTemplateAttributeExclusion(models.Model):
     _inherit = ['product.template.attribute.exclusion', 'pos.load.mixin']
 
     @api.model
-    def _load_pos_data_domain(self, data):
+    def _load_pos_data_domain(self, data, config):
         loaded_product_tmpl_ids = list({p['id'] for p in data['product.template']})
         return [('product_tmpl_id', 'in', loaded_product_tmpl_ids)]
 
     @api.model
-    def _load_pos_data_fields(self, config_id):
+    def _load_pos_data_fields(self, config):
         return ['value_ids']

--- a/addons/point_of_sale/models/product_category.py
+++ b/addons/point_of_sale/models/product_category.py
@@ -7,5 +7,5 @@ class ProductCategory(models.Model):
     _inherit = ['product.category', 'pos.load.mixin']
 
     @api.model
-    def _load_pos_data_fields(self, config_id):
+    def _load_pos_data_fields(self, config):
         return ['id', 'name', 'parent_id']

--- a/addons/point_of_sale/models/product_combo.py
+++ b/addons/point_of_sale/models/product_combo.py
@@ -12,11 +12,11 @@ class ProductCombo(models.Model):
     qty_free = fields.Integer(string="Free quantity", default=1, help="Number of free items included in the combo.")
 
     @api.model
-    def _load_pos_data_domain(self, data):
+    def _load_pos_data_domain(self, data, config):
         return [('id', 'in', list(set().union(*[product.get('combo_ids') for product in data['product.template']])))]
 
     @api.model
-    def _load_pos_data_fields(self, config_id):
+    def _load_pos_data_fields(self, config):
         return ['id', 'name', 'combo_item_ids', 'base_price', 'qty_free', 'qty_max']
 
     @api.constrains('qty_max')

--- a/addons/point_of_sale/models/product_combo_item.py
+++ b/addons/point_of_sale/models/product_combo_item.py
@@ -8,9 +8,9 @@ class ProductComboItem(models.Model):
     _inherit = ['product.combo.item', 'pos.load.mixin']
 
     @api.model
-    def _load_pos_data_domain(self, data):
+    def _load_pos_data_domain(self, data, config):
         return [('id', 'in', list(set().union(*[combo.get('combo_item_ids') for combo in data['product.combo']])))]
 
     @api.model
-    def _load_pos_data_fields(self, config_id):
+    def _load_pos_data_fields(self, config):
         return ['id', 'combo_id', 'product_id', 'extra_price']

--- a/addons/point_of_sale/models/product_pricelist.py
+++ b/addons/point_of_sale/models/product_pricelist.py
@@ -7,13 +7,12 @@ class ProductPricelist(models.Model):
     _inherit = ['product.pricelist', 'pos.load.mixin']
 
     @api.model
-    def _load_pos_data_domain(self, data):
-        config_id = self.env['pos.config'].browse(data['pos.config'][0]['id'])
+    def _load_pos_data_domain(self, data, config):
         pricelist_ids = [preset['pricelist_id'] for preset in data['pos.preset']]
-        return [('id', 'in', config_id._get_available_pricelists().ids + pricelist_ids)]
+        return [('id', 'in', config._get_available_pricelists().ids + pricelist_ids)]
 
     @api.model
-    def _load_pos_data_fields(self, config_id):
+    def _load_pos_data_fields(self, config):
         return ['id', 'name', 'display_name', 'item_ids']
 
 
@@ -22,7 +21,7 @@ class ProductPricelistItem(models.Model):
     _inherit = ['product.pricelist.item', 'pos.load.mixin']
 
     @api.model
-    def _load_pos_data_domain(self, data):
+    def _load_pos_data_domain(self, data, config):
         product_tmpl_ids = [p['product_tmpl_id'] for p in data['product.product']]
         product_ids = [p['id'] for p in data['product.product']]
         product_categ = [c['id'] for c in data['product.category']]
@@ -38,7 +37,7 @@ class ProductPricelistItem(models.Model):
         ]
 
     @api.model
-    def _load_pos_data_fields(self, config_id):
+    def _load_pos_data_fields(self, config):
         return ['product_tmpl_id', 'product_id', 'pricelist_id', 'price_surcharge', 'price_discount', 'price_round',
                 'price_min_margin', 'price_max_margin', 'company_id', 'currency_id', 'date_start', 'date_end', 'compute_price',
                 'fixed_price', 'percent_price', 'base_pricelist_id', 'base', 'categ_id', 'min_quantity']

--- a/addons/point_of_sale/models/product_product.py
+++ b/addons/point_of_sale/models/product_product.py
@@ -8,12 +8,11 @@ class ProductProduct(models.Model):
     _inherit = ['product.product', 'pos.load.mixin']
 
     @api.model
-    def _load_pos_data_domain(self, data):
+    def _load_pos_data_domain(self, data, config):
         return [('product_tmpl_id', 'in', [p['id'] for p in data['product.template']])]
 
     @api.model
-    def _load_pos_data_fields(self, config_id):
-        config = self.env['pos.config'].browse(config_id)
+    def _load_pos_data_fields(self, config):
         taxes = self.env['account.tax'].search(self.env['account.tax']._check_company_domain(config.company_id.id))
         product_fields = taxes._eval_taxes_computation_prepare_product_fields()
         return list(product_fields.union({
@@ -31,13 +30,16 @@ class ProductProduct(models.Model):
                     "Deleting a product available in a session would be like attempting to snatch a hamburger from a customer’s hand mid-bite; chaos will ensue as ketchup and mayo go flying everywhere!",
                 ))
 
-    def _post_read_pos_data(self, data):
-        config = self.env['pos.config'].browse(self.env.context.get('config_id'))
+    @api.model
+    def _load_pos_data_read(self, records, config):
+        read_records = super()._load_pos_data_read(records, config)
         different_currency = config.currency_id != self.env.company.currency_id
         if different_currency:
-            for product in data:
-                product['lst_price'] = self.env.company.currency_id._convert(product['lst_price'], config.currency_id, self.env.company, fields.Date.today())
-        return super()._post_read_pos_data(data)
+            for product in read_records:
+                product['lst_price'] = config.currency_id._convert(
+                    product['lst_price'], self.env.company.currency_id, self.env.company, fields.Date.today()
+                )
+        return read_records
 
     def _can_return_content(self, field_name=None, access_token=None):
         if field_name == "image_128" and self.sudo().available_in_pos:

--- a/addons/point_of_sale/models/product_tag.py
+++ b/addons/point_of_sale/models/product_tag.py
@@ -11,7 +11,7 @@ class ProductTag(models.Model):
     has_image = fields.Boolean(compute='_compute_has_image')
 
     @api.model
-    def _load_pos_data_fields(self, config_id):
+    def _load_pos_data_fields(self, config):
         return ['name', 'pos_description', 'color', 'has_image', 'write_date']
 
     @api.depends('has_image')

--- a/addons/point_of_sale/models/product_template.py
+++ b/addons/point_of_sale/models/product_template.py
@@ -63,24 +63,22 @@ class ProductTemplate(models.Model):
     def create_product_variant_from_pos(self, attribute_value_ids, config_id):
         """ Create a product variant from the POS interface. """
         self.ensure_one()
+        pos_config = self.env['pos.config'].browse(config_id)
         product_template_attribute_value_ids = self.env['product.template.attribute.value'].browse(attribute_value_ids)
         product_variant = self._create_product_variant(product_template_attribute_value_ids)
         return {
-            'product.product': product_variant.read(self.env['product.product']._load_pos_data_fields(config_id), load=False),
+            'product.product': product_variant.read(self.env['product.product']._load_pos_data_fields(pos_config), load=False),
         }
 
     @api.model
-    def _load_pos_data_domain(self, data):
+    def _load_pos_data_domain(self, data, config):
         domain = [
-            *self.env['product.template']._check_company_domain(data['pos.config'][0]['company_id']),
+            *self.env['product.template']._check_company_domain(config.company_id),
             ('available_in_pos', '=', True),
             ('sale_ok', '=', True),
         ]
-        limited_categories = data['pos.config'][0]['limit_categories']
-        if limited_categories:
-            available_category_ids = data['pos.config'][0]['iface_available_categ_ids']
-            category_ids = self.env['pos.category'].browse(available_category_ids)._get_descendants().ids
-            domain += [('pos_categ_ids', 'in', category_ids)]
+        if config.limit_categories:
+            domain += [('pos_categ_ids', 'in', config.iface_available_categ_ids.ids)]
         return domain
 
     @api.model
@@ -95,11 +93,11 @@ class ProductTemplate(models.Model):
                 product_tmpls += product_tmpl.combo_ids.combo_item_ids.product_id.product_tmpl_id
 
         combo_domain = [('id', 'in', product_tmpls.combo_ids.ids)]
-        combo_fields = self.env['product.combo']._load_pos_data_fields(config.id)
-        combo_read = self.env['product.combo'].search_read(combo_domain, combo_fields, load=False)
+        combo_records = self.env['product.combo'].search(combo_domain)
+        combo_read = self.env['product.combo']._load_pos_data_read(combo_records, config)
         combo_item_domain = [('combo_id', 'in', product_tmpls.combo_ids.ids)]
-        combo_item_fields = self.env['product.combo.item']._load_pos_data_fields(config.id)
-        combo_item_read = self.env['product.combo.item'].search_read(combo_item_domain, combo_item_fields, load=False)
+        combo_item_records = self.env['product.combo.item'].search(combo_item_domain)
+        combo_item_read = self.env['product.combo.item']._load_pos_data_read(combo_item_records, config)
 
         products = product_tmpls.product_variant_ids
 
@@ -112,27 +110,22 @@ class ProductTemplate(models.Model):
 
         # product.template.attribute.value & product.template.attribute.line loading
         product_tmpl_attr_line = product_tmpls.attribute_line_ids
+        product_tmpl_attr_line_read = product_tmpl_attr_line._load_pos_data_read(product_tmpl_attr_line, config)
         product_tmpl_attr_value = product_tmpls.attribute_line_ids.product_template_value_ids
-        product_tmpl_attr_value_fields = product_tmpl_attr_value._load_pos_data_fields(config.id)
-        product_tmpl_attr_line_fields = product_tmpl_attr_line._load_pos_data_fields(config.id)
-        product_tmpl_attr_value_read = product_tmpl_attr_value.read(product_tmpl_attr_value_fields, load=False)
-        product_tmpl_attr_line_read = product_tmpl_attr_line.read(product_tmpl_attr_line_fields, load=False)
+        product_tmpl_attr_value_read = product_tmpl_attr_value._load_pos_data_read(product_tmpl_attr_value, config)
 
         # product.template.attribute.exclusion loading
         product_tmpl_excl = self.env['product.template.attribute.exclusion']
         product_tmpl_exclusion = product_tmpl_attr_value.exclude_for + product_tmpl_excl.search([
             ('product_tmpl_id', 'in', product_tmpls.ids),
         ])
-        product_tmpl_exclusion_fields = product_tmpl_exclusion._load_pos_data_fields(config.id)
-        product_tmpl_exclusion_read = product_tmpl_exclusion.read(product_tmpl_exclusion_fields, load=False)
+        product_tmpl_exclusion_read = product_tmpl_excl._load_pos_data_read(product_tmpl_exclusion, config)
 
         # product.product loading
-        product_fields = products._load_pos_data_fields(config.id)
-        product_read = products.with_context(display_default_code=False).read(product_fields, load=False)
+        product_read = products._load_pos_data_read(products.with_context(display_default_code=False), config)
 
         # product.template loading
-        product_tmpl_fields = self._load_pos_data_fields(config.id)
-        product_tmpl_read = product_tmpls.read(product_tmpl_fields, load=False)
+        product_tmpl_read = self._load_pos_data_read(product_tmpls, config)
 
         # product.uom loading
         packaging_domain = Domain([('product_id', 'in', products.ids)])
@@ -144,22 +137,23 @@ class ProductTemplate(models.Model):
             flat = [item for sublist in barcode for item in sublist]
             packaging_domain = expression.OR([packaging_domain, [('barcode', 'in', flat)]])
 
-        packaging_fields = self.env['product.uom']._load_pos_data_fields(config.id)
-        packaging = self.env['product.uom'].search(packaging_domain)
+        product_uom = self.env['product.uom']
+        packaging = product_uom.search(packaging_domain)
         condition = packaging and packaging.product_id
+        packaging_read = product_uom._load_pos_data_read(packaging, config) if condition else []
 
         # account.tax loading
-        tax_domain = self.env['account.tax']._check_company_domain(config.company_id.id)
+        account_tax = self.env['account.tax']
+        tax_domain = account_tax._check_company_domain(config.company_id.id)
         tax_domain = expression.AND([tax_domain, [['id', 'in', product_tmpls.taxes_id.ids]]])
-        tax_fields = self.env['account.tax']._load_pos_data_fields(config.id)
-        tax_read = self.env['account.tax'].search_read(tax_domain, tax_fields, load=False)
+        tax_read = account_tax._load_pos_data_read(account_tax.search(tax_domain), config)
 
         return {
             **pricelists,
             'account.tax': tax_read,
-            'product.product': products.with_context(config_id=config_id)._post_read_pos_data(product_read),
-            'product.template': self.with_context(config_id=config.id)._post_read_pos_data(product_tmpl_read),
-            'product.uom': packaging.read(packaging_fields, load=False) if condition else [],
+            'product.product': product_read,
+            'product.template': product_tmpl_read,
+            'product.uom': packaging_read,
             'product.combo': combo_read,
             'product.combo.item': combo_item_read,
             'product.template.attribute.value': product_tmpl_attr_value_read,
@@ -176,12 +170,12 @@ class ProductTemplate(models.Model):
             'pos_optional_product_ids', 'sequence', 'product_tag_ids'
         ]
 
-    def _load_pos_data(self, data):
-        config = self.env['pos.config'].browse(data['pos.config'][0]['id'])
+    @api.model
+    def _load_pos_data_search_read(self, data, config):
         limit_count = config.get_limited_product_count()
         pos_limited_loading = self.env.context.get('pos_limited_loading', True)
         if limit_count and pos_limited_loading:
-            query = self._where_calc(self._load_pos_data_domain(data))
+            query = self._where_calc(self._load_pos_data_domain(data, config))
             sql = SQL(
                 """
                     WITH pm AS (
@@ -208,7 +202,7 @@ class ProductTemplate(models.Model):
             product_tmpl_ids = [r[0] for r in self.env.execute_query(sql)]
             products = self._load_product_with_domain([('id', 'in', product_tmpl_ids)])
         else:
-            domain = self._load_pos_data_domain(data)
+            domain = self._load_pos_data_domain(data, config)
             products = self._load_product_with_domain(domain)
 
         product_combo = products.filtered(lambda p: p['type'] == 'combo')
@@ -228,13 +222,13 @@ class ProductTemplate(models.Model):
         if products.filtered(lambda p: p.pos_optional_product_ids):
             products |= products.mapped("pos_optional_product_ids")
 
-        fields = self._load_pos_data_fields(config.id)
-        return self.with_context(config_id=config.id)._post_read_pos_data(products.read(fields, load=False))
+        return self._load_pos_data_read(products, config)
 
-    def _post_read_pos_data(self, data):
-        config = self.env['pos.config'].browse(self.env.context.get('config_id'))
-        self._process_pos_ui_product_product(data, config)
-        return super()._post_read_pos_data(data)
+    @api.model
+    def _load_pos_data_read(self, records, config):
+        read_records = super()._load_pos_data_read(records, config)
+        self._process_pos_ui_product_product(read_records, config)
+        return read_records
 
     def _load_product_with_domain(self, domain, load_archived=False, offset=0, limit=0):
         context = {**self.env.context, 'display_default_code': False, 'active_test': not load_archived}

--- a/addons/point_of_sale/models/product_uom.py
+++ b/addons/point_of_sale/models/product_uom.py
@@ -7,5 +7,5 @@ class ProductUom(models.Model):
     _inherit = ['product.uom', 'pos.load.mixin']
 
     @api.model
-    def _load_pos_data_fields(self, config_id):
+    def _load_pos_data_fields(self, config):
         return ['id', 'barcode', 'product_id', 'uom_id']

--- a/addons/point_of_sale/models/res_company.py
+++ b/addons/point_of_sale/models/res_company.py
@@ -31,11 +31,11 @@ class ResCompany(models.Model):
         required=True)
 
     @api.model
-    def _load_pos_data_domain(self, data):
-        return [('id', '=', data['pos.config'][0]['company_id'])]
+    def _load_pos_data_domain(self, data, config):
+        return [('id', '=', config.company_id.id)]
 
     @api.model
-    def _load_pos_data_fields(self, config_id):
+    def _load_pos_data_fields(self, config):
         return [
             'id', 'currency_id', 'email', 'website', 'company_registry', 'vat', 'name', 'phone', 'partner_id',
             'country_id', 'state_id', 'tax_calculation_rounding_method', 'nomenclature_id', 'point_of_sale_use_ticket_qr_code',

--- a/addons/point_of_sale/models/res_country.py
+++ b/addons/point_of_sale/models/res_country.py
@@ -6,5 +6,5 @@ class ResCountry(models.Model):
     _inherit = ['res.country', 'pos.load.mixin']
 
     @api.model
-    def _load_pos_data_fields(self, config_id):
+    def _load_pos_data_fields(self, config):
         return ['id', 'name', 'code', 'vat_label']

--- a/addons/point_of_sale/models/res_country_state.py
+++ b/addons/point_of_sale/models/res_country_state.py
@@ -6,5 +6,5 @@ class ResCountryState(models.Model):
     _inherit = ['res.country.state', 'pos.load.mixin']
 
     @api.model
-    def _load_pos_data_fields(self, config_id):
+    def _load_pos_data_fields(self, config):
         return ['id', 'name', 'code', 'country_id']

--- a/addons/point_of_sale/models/res_currency.py
+++ b/addons/point_of_sale/models/res_currency.py
@@ -6,12 +6,13 @@ class ResCurrency(models.Model):
     _inherit = ['res.currency', 'pos.load.mixin']
 
     @api.model
-    def _load_pos_data_domain(self, data):
-        company_currency_id = self.env['res.company'].browse(data['pos.config'][0]['company_id']).currency_id.id
-        if company_currency_id != data['pos.config'][0]['currency_id']:
-            return [('id', 'in', [company_currency_id, data['pos.config'][0]['currency_id']])]
-        return [('id', '=', data['pos.config'][0]['currency_id'])]
+    def _load_pos_data_domain(self, data, config):
+        company_currency_id = config.company_id.currency_id.id
+        config_currency_id = config.currency_id.id
+        if company_currency_id != config_currency_id:
+            return [('id', 'in', [company_currency_id, config_currency_id])]
+        return [('id', '=', config_currency_id)]
 
     @api.model
-    def _load_pos_data_fields(self, config_id):
+    def _load_pos_data_fields(self, config):
         return ['id', 'name', 'symbol', 'position', 'rounding', 'rate', 'decimal_places', 'iso_numeric']

--- a/addons/point_of_sale/models/res_lang.py
+++ b/addons/point_of_sale/models/res_lang.py
@@ -6,5 +6,5 @@ class ResLang(models.Model):
     _inherit = ['res.lang', 'pos.load.mixin']
 
     @api.model
-    def _load_pos_data_fields(self, config_id):
+    def _load_pos_data_fields(self, config):
         return ['id', 'name', 'code', 'flag_image_url', 'display_name']

--- a/addons/point_of_sale/models/res_partner.py
+++ b/addons/point_of_sale/models/res_partner.py
@@ -48,19 +48,17 @@ class ResPartner(models.Model):
             new_partners = self.search(domain, offset=offset, limit=100)
         fiscal_positions = new_partners.fiscal_position_id
         return {
-            'res.partner': new_partners.read(self._load_pos_data_fields(config_id), load=False),
-            'account.fiscal.position': fiscal_positions.read(self.env['account.fiscal.position']._load_pos_data_fields(config_id), load=False)
+            'res.partner': self._load_pos_data_read(new_partners, config),
+            'account.fiscal.position': self.env['account.fiscal.position']._load_pos_data_read(fiscal_positions, config),
         }
 
     @api.model
-    def _load_pos_data_domain(self, data):
-        config_id = self.env['pos.config'].browse(data['pos.config'][0]['id'])
-
+    def _load_pos_data_domain(self, data, config):
         # Collect partner IDs from loaded orders
         loaded_order_partner_ids = {order['partner_id'] for order in data['pos.order']}
 
         # Extract partner IDs from the tuples returned by get_limited_partners_loading
-        limited_partner_ids = {partner[0] for partner in config_id.get_limited_partners_loading()}
+        limited_partner_ids = {partner[0] for partner in config.get_limited_partners_loading()}
 
         limited_partner_ids.add(self.env.user.partner_id.id)  # Ensure current user is included
         partner_ids = limited_partner_ids.union(loaded_order_partner_ids)
@@ -71,7 +69,7 @@ class ResPartner(models.Model):
             partner.fiscal_position_id = self.env['account.fiscal.position'].with_company(self.env.company)._get_fiscal_position(partner)
 
     @api.model
-    def _load_pos_data_fields(self, config_id):
+    def _load_pos_data_fields(self, config):
         return [
             'id', 'name', 'street', 'street2', 'city', 'state_id', 'country_id', 'vat', 'lang', 'phone', 'zip', 'email',
             'barcode', 'write_date', 'property_product_pricelist', 'parent_name', 'pos_contact_address',

--- a/addons/point_of_sale/models/res_users.py
+++ b/addons/point_of_sale/models/res_users.py
@@ -6,16 +6,17 @@ class ResUsers(models.Model):
     _inherit = ['res.users', 'pos.load.mixin']
 
     @api.model
-    def _load_pos_data_domain(self, data):
+    def _load_pos_data_domain(self, data, config):
         return [('id', '=', self.env.uid)]
 
     @api.model
-    def _load_pos_data_fields(self, config_id):
+    def _load_pos_data_fields(self, config):
         return ['id', 'name', 'partner_id', 'all_group_ids']
 
-    def _post_read_pos_data(self, data):
-        config_id = self.env['pos.config'].browse(self.env.context.get('config_id'))
-        if data:
-            data[0]['role'] = 'manager' if config_id.group_pos_manager_id.id in data[0]['all_group_ids'] else 'cashier'
-            del data[0]['all_group_ids']
-        return super()._post_read_pos_data(data)
+    @api.model
+    def _load_pos_data_read(self, records, config):
+        read_records = super()._load_pos_data_read(records, config)
+        if read_records:
+            read_records[0]['role'] = 'manager' if config.group_pos_manager_id.id in read_records[0]['all_group_ids'] else 'cashier'
+            del read_records[0]['all_group_ids']
+        return read_records

--- a/addons/point_of_sale/models/resource_calendar_attendance.py
+++ b/addons/point_of_sale/models/resource_calendar_attendance.py
@@ -6,12 +6,12 @@ class ResourceCalendarAttendance(models.Model):
     _inherit = ['resource.calendar.attendance', 'pos.load.mixin']
 
     @api.model
-    def _load_pos_data_domain(self, data):
+    def _load_pos_data_domain(self, data, config):
         attendance_ids = []
         for preset in data['pos.preset']:
             attendance_ids += preset['attendance_ids']
         return [('id', 'in', attendance_ids)]
 
     @api.model
-    def _load_pos_data_fields(self, config_id):
+    def _load_pos_data_fields(self, config):
         return ['id', 'hour_from', 'hour_to', 'dayofweek', 'day_period']

--- a/addons/point_of_sale/models/stock_picking.py
+++ b/addons/point_of_sale/models/stock_picking.py
@@ -183,11 +183,11 @@ class StockPickingType(models.Model):
                 raise ValidationError(_("You cannot archive '%(picking_type)s' as it is used by POS configuration '%(config)s'.", picking_type=picking_type.name, config=pos_config.name))
 
     @api.model
-    def _load_pos_data_domain(self, data):
-        return [('id', '=', data['pos.config'][0]['picking_type_id'])]
+    def _load_pos_data_domain(self, data, config):
+        return [('id', '=', config.picking_type_id.id)]
 
     @api.model
-    def _load_pos_data_fields(self, config_id):
+    def _load_pos_data_fields(self, config):
         return ['id', 'use_create_lots', 'use_existing_lots']
 
 

--- a/addons/point_of_sale/models/uom.py
+++ b/addons/point_of_sale/models/uom.py
@@ -8,5 +8,5 @@ class UomUom(models.Model):
     is_pos_groupable = fields.Boolean(string='Group Products in POS', help="Check if you want to group products of this unit in point of sale orders")
 
     @api.model
-    def _load_pos_data_fields(self, config_id):
+    def _load_pos_data_fields(self, config):
         return ['id', 'name', 'factor', 'is_pos_groupable', 'parent_path', 'rounding']
